Resolve Log authorization detail failure message by override ToString in each Requirement

wu-yafeng opened the pull request:
Log authorization detail failure message by override ToString in each Requirement
 - Add AuthorizationFailure parameter to internal LoggingExtensions.UserAuthorizationFailed
 - Override each Requirement ToString method

Failed Authorization Policy Details in Logs #7789

Commit:
Loop is not neccessary for checking createSelectorForSilentRouteProviders when first routeProvider is not IActionHttpMethodProvider

Commit:
Merge branch 'master' of https://github.com/aspnet/AspNetCore

Commit:
#7789 Failed Authorization Policy Details in Logs

Commit:
Merge remote-tracking branch 'aspnet/master'

HaoK commented:
@wu-yafeng this looks like a reasonable start, do you think you could add some tests to demonstrate what the log output would look like?

wu-yafeng commented:
> @wu-yafeng this looks like a reasonable start, do you think you could add some tests to demonstrate what the log output would look like?

sure

Commit:
Add tests to demonstrate the log output.
- Fix requirement ToString format
- Add requirement ToString tests
- Add Authorization log output tests

Commit:
Fix CI Failure

anurse commented:
@HaoK 	if we can get this reviewed an in for preview8, awesome. Otherwise it will go to 3.1

HaoK commented on src/Security/Authorization/Core/src/LoggingExtensions.cs:
```diff
@@ -16,16 +17,22 @@ static LoggingExtensions()
+        public static void UserAuthorizationFailed(this ILogger logger, AuthorizationFailure failure)
+        {
+            var reason = failure.FailCalled
+                ? "Explicit called fail"
```
nit: "Fail() was explicitly called."

HaoK commented on src/Security/Authorization/Core/src/LoggingExtensions.cs:
```diff
@@ -16,16 +17,22 @@ static LoggingExtensions()
+        {
+            var reason = failure.FailCalled
+                ? "Explicit called fail"
+                : "Following requirement not met " + Environment.NewLine + string.Join(Environment.NewLine, failure.FailedRequirements);
```
nit: "These requirements were not met:"

HaoK commented on src/Security/Authorization/Core/src/DenyAnonymousAuthorizationRequirement.cs:
```diff
@@ -29,5 +29,10 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
+
+        public override string ToString()
+        {
+            return $"{nameof(DenyAnonymousAuthorizationRequirement)}:User authenticationted";
```
type: "User authenticated"?

HaoK commented on src/Security/Authorization/Core/src/AssertionRequirement.cs:
```diff
@@ -56,5 +56,10 @@ public async Task HandleAsync(AuthorizationHandlerContext context)
+
+        public override string ToString()
+        {
+            return $"{nameof(AssertionRequirement)}:{nameof(Handler)} return {bool.TrueString}";
```
nit: "{nameof(Handler)} assertion should evaluate to true."

HaoK commented on src/Security/Authorization/Core/src/ClaimsAuthorizationRequirement.cs:
```diff
@@ -69,5 +70,14 @@ protected override Task HandleRequirementAsync(AuthorizationHandlerContext conte
+        {
+            var value = (AllowedValues == null || !AllowedValues.Any())
+                ? string.Empty
+                : $" AND Claim.Value in ({string.Join("|", AllowedValues)})";
```
nit: " and Claim.Value is one of the following values: ({string.Join("|", AllowedValues)})"

HaoK reviewed (changes requested):
Looks good except for some minor text changes

wu-yafeng commented:
Poor english 😢 ,I'll fix it

Commit:
Fix requirement text
- Fix `AssertRequirement.ToString` text and test case.
- Fix `DenyAnonymous.ToString` misspell and test case.
- Fix `ClaimsAuthorizationRequirement` text and tes case
- Update DefaultAuthorizationServiceTests
- Update LoggingExtensions log message when authorize fail

anurse commented:
@HaoK let's try to close this out now that we're in to the free-for-all 5.0 timeframe ;)

HaoK committed:
Add whitespace

HaoK committed:
Update whitespace

HaoK commented:
Thanks @wu-yafeng I had to clean up some minor things and rebased, and merged these changes in PR https://github.com/aspnet/AspNetCore/pull/15350
